_specifications/specification-3-17.md

```diff
@@ -8217,6 +8217,182 @@ export interface Moniker {
 
 Server implementations of this method should ensure that the moniker calculation matches to those used in the corresponding LSIF implementation to ensure symbols can be associated correctly across IDE sessions and LSIF indexes.
 
+#### <a href="#textDocument_inlayHints" name="textDocument_inlayHints" class="anchor">Inlay hints (:leftwards_arrow_with_hook:)</a>
+
+> *Since version 3.17.0*
+
+Inlay hints are short textual annotations that are attached to ranges in the source code.
+
+**TODO**: Is "inlay hints" too presentational a name? The protocol allows other UI (e.g. pop-up when cursor is in the target range).
+
+These typically spell out some inferred information, such as the parameter name when passing a value to a function.
+
+```typescript
+/**
+ * Well-known kinds of information conveyed by InlayHints.
+ * This is not an exhaustive list, servers may use other categories as needed.
+ */
+export enum InlayHintCategory {
```

Comment:
Are those categories necessary for a 1st iteration? I don't see categories in "Hover" nor "CodeLens" for instance. I think a 1st iteration should take place without them.

Reply:
> I find the configurability argument quite compelling, as this is a highly invasive/visible feature.

The thing is that to properly handle it, it would require tools to dynamically generate some page which lists the available categories in the LS and allow to enable/disable some of them. The LS would require to allow more introspective than usual; categories/kinds are usually hardcoded pre-identified enums of things shared by many languages more than dynamically filled placeholders for LS-specific stuff.
If we need to distinguish different kinds, I think it's better to do it just like Completion and other do: the protocol defines a static list of supported kinds; that clients can then code a behavior for.

Reply:
Oh, that was definitely the intention here: clients could show fixed checkboxes for Params/Types/Other.
(In fact I thought about using classes for configuration, but it seems too complicated for exactly the reason you mention)

The reason to allow other values is just to make protocol versioning/extensions easier:
 - avoid deadlock where neither clients/servers/spec feel they should be the first to support a new category
 - avoid complicated capability negotiation (completionItemKind.valueSet) to ensure back-compatibility
It's possible different servers use the same non-standard category name to mean different things, I don't think this is a large risk.

Reply:
(I've tweaked some comments and such, trying to make this clearer)